src/backend/cdb/cdbmutate.c

```diff
@@ -2610,12 +2619,115 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
 	return true;
 }
 
+static bool
+root_slice_is_executed_on_coordinator(Plan *plan, PlannerInfo *root)
+{
+	Query	   *query = root->parse;
+
+	if (query->parentStmtType != PARENTSTMTTYPE_NONE)
+	{
+		/*
+		 * For CTAS, SELECT INTO, COPY INTO, REFRESH MATERIALIZED VIEW, the
+		 * root slice is not executed on the coordinator.
+		 */
+		return false;
+	}
+	else if (IsA(plan, ModifyTable))
+	{
+		ModifyTable *mt = (ModifyTable *) plan;
+
+		if (list_length(mt->resultRelations) > 0)
+		{
+			ListCell   *lc = list_head(mt->resultRelations);
+			Oid			reloid = getrelid(lfirst_int(lc), query->rtable);
+
+			if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
```

Comment:
It seems we can heavily simplify this conditions.
```suggestion
			return (GpPolicyFetch(reloid)->ptype == POLICYTYPE_ENTRY);
```

Reply:
> It seems we can heavily simplify this conditions.

I removed the else, so now it can't be so simplified.

Reply:
Sometimes I think you joking me.

What was the reason of removal?

Reply:
> What was the reason of removal?

I revised the conditions in the executor.

Reply:
Please give a link or explain it somehow.

Reply:
> Please give a link or explain it somehow.

I tried to describe it in the header:

> 1) For `CTAS`, `SELECT INTO`, `COPY INTO`, `REFRESH MATERIALIZED VIEW`, [the root `slice` is not executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5231-L5247).
> 2) For the modification node, the table distribution policy is checked, and if the table is not master-only, then [the root `slice` is not executed on the `coordinator`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L5051-L5116).
> 3) Parallel cursors can only be generated by the postgres planner, which means [it will set the correct `flow`](https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L1809-L1843).
If it is not a `singleton`, then the root `slice` is not executed on the `coordinator`.
If this is a `singleton`, but the `locus` is not `Entry`, not `General`, and not `SingleQE`, then the root `slice` is also not executed on the `coordinator`.
> 4) The postgres planner fills `flow` correctly.
If it is not a `singleton`, then the root `slice` is not executed on the `coordinator`.
If this is a `singleton` on a segment, then the root `slice` is also not executed on the `coordinator`.
> 5) In all other cases, the root `slice` is executed on the `coordinator`.

Reply:
I don't see any explanation here about why this conditions (or their subset) can't use `else if`.

Reply:
> I don't see any explanation here about why this conditions (or their subset) can't use `else if`.

By default, the root slice is executed on the coordinator, and only under certain conditions is it not.

Reply:
> By default, the root slice is executed on the coordinator, and only under certain conditions is it not.

How this explains my question about `else` operator?

Reply:
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2633-R2636) the else operator is not needed

Reply:
Can you clarify why it was necessary to remove else condition in this line?

Reply:
Because if `if` ends with a `return`, then the `else` is redundant.

Reply:
True.
What about other cases? Can you explain them in their threads?

Reply:
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2655-R2657) we can use `else`.

Reply:
> True. What about other cases? Can you explain them in their threads?

I'm thinking about it...